### PR TITLE
Simplified generics implementation

### PR DESCRIPTION
**Pull request by AriFordsham**
The main advantage, besides simplified implementation, Is that the Plutarch-side code `[[PType]]` is exposed via the constraint `IsPCodeOf`, rather than being an existential quantification, vastly improving ergonomics.

### CODE REVIEW

File: Plutarch/Core.hs [hunk 1]
```diff
@@ -1,6 +1,7 @@
 {-# LANGUAGE FlexibleInstances #-}
 {-# LANGUAGE UndecidableInstances #-}
 {-# LANGUAGE UndecidableSuperClasses #-}
+{-# LANGUAGE PartialTypeSignatures #-}
```
Comment (L-as) [inline Plutarch/Core.hs:4]: Do we need this?

File: Plutarch/Core.hs [hunk 2]
```diff
@@ -306,101 +307,21 @@ class PDSL edsl => PAp (f :: Type -> Type) edsl where
...
-            PIsSumR
-              { inner = Proxy @(innerh : innert)
-              , to = \case
-                  SOP.SOP (SOP.Z x) -> SOP.SOP $ SOP.Z $ toh x
-                  SOP.SOP (SOP.S x) -> case tot $ SOP.SOP $ x of SOP.SOP y -> SOP.SOP (SOP.S y)
-              , from = \case
-                  SOP.SOP (SOP.Z x) -> SOP.SOP $ SOP.Z $ fromh x
-                  SOP.SOP (SOP.S x) -> case fromt $ SOP.SOP $ x of SOP.SOP y -> SOP.SOP (SOP.S y)
-              }
+class (forall a. t ~ Term edsl a) => PIsTerm edsl t
```
Comment (L-as) [inline Plutarch/Core.hs:310]: This doesn't make sense to me. How can this ever be the case?

File: Plutarch/Core.hs [hunk 3]
```diff
@@ -306,101 +307,21 @@ class PDSL edsl => PAp (f :: Type -> Type) edsl where
...
-  , PIsSum edsl (SOPG.GCode (PConcrete edsl a))
-  , PReprSort a ~ PReprSOP
-  ) =>
-  PIsSOP (edsl :: PDSLKind) (a :: PType)
-  where
-  esop :: Proxy edsl -> Proxy a -> PIsSumR edsl (SOPG.GCode (PConcrete edsl a))
-instance
-  ( PGeneric a
-  , PIsSum edsl (SOPG.GCode (PConcrete edsl a))
+  , PIsSum edsl (SOP.Code (PConcrete edsl a))
```
Comment (L-as) [inline Plutarch/Core.hs:321]: This seems like a mistaken change? You probably still want `GCode`.

**Review by L-as — COMMENTED** [review]
Thanks!

File: Plutarch/Core.hs (hunk 1 above)
Comment (AriFordsham) [reply Plutarch/Core.hs:4]: Oops... mistake

File: Plutarch/Core.hs (hunk 3 above)
Comment (AriFordsham) [reply Plutarch/Core.hs:321]: `SOP.Code` is a synonym. Nevertheless, using in place of `GCode` has caused me problems, so I'm adding it back.

File: Plutarch/Core.hs (hunk 2 above)
Comment (AriFordsham) [reply Plutarch/Core.hs:310]: The intention is 'for some `a`'. I thought that was correct, how is it done?

Comment (Icelandjack) [reply Plutarch/Core.hs:310]: There are no existential constraints but for equality you can write `Split t ~ Term edsl` instead

```haskell
type
  Split :: k -> (j -> k)
type family
  Split a where
  Split (f _) = f
```

File: Plutarch/Core.hs (hunk 3 above)
Comment (L-as) [reply Plutarch/Core.hs]: `SOP.Code` is not a synonym.

Comment (AriFordsham) [reply Plutarch/Core.hs]: 😮 This led me astray: https://hackage.haskell.org/package/generics-sop-0.5.1.2/docs/Generics-SOP.html#t:Code:~:text=type%20Code%20a%20%3D%20GCode%20a

File: Plutarch/Core.hs [hunk 4]
```diff
@@ -411,6 +341,31 @@ type PSOP edsl =
   , IsPType edsl PPType
   )
 
+class
+  ( SOP.AllZip2 (SOP.LiftedCoercible SOP.I (Term edsl)) tss pss
+  , SOP.AllZip2 (SOP.LiftedCoercible (Term edsl) SOP.I) pss tss
+  ) =>
+  IsPCodeOf edsl pss tss
```
Comment (re-xyr) [inline Plutarch/Core.hs:348]: Could we have `| edsl pss -> tss, edsl tss -> pss`?

Comment (AriFordsham) [reply Plutarch/Core.hs:348]: Indeed we cannot, which is unfortunate. It will not allow the 'catch-all' instance of the constraint newtype pattern.

**Review by L-as — COMMENTED** [review]
This doesn't actually work because you can't put this into the `PSOP` constraint because there's no way to compute the `pss` from the `tss`.

**Comment by L-as** [comment]
Using `hcoerce` is a minor upgrade, but actually, it seems it's also implemented under the hood with `unsafeCoerce`... Less `unsafeCoerce` is still better, but right now we don't use any `unsafeCoerce`.

**Comment by AriFordsham** [comment]
`hcoerce` narrows the unsafety surface of coercion by using a specialization expected to be safe. This could be implemented fully safely, at the cost of (compile-time) performance.

File: Plutarch/Core.hs (hunk 4 above)
Comment (AriFordsham) [reply Plutarch/Core.hs:348]: https://github.com/Plutonomicon/plutarch-core/pull/18#pullrequestreview-1111511619

Comment (AriFordsham) [reply Plutarch/Core.hs:348]: It actually may, if we use a type family to compute `pss` from `tss`.

**Comment by re-xyr** [comment]
I experimented a little and got a version with associated type families and got this to work:

```hs
class
  ( SOP.AllZip (SOP.LiftedCoercible SOP.I (Term edsl)) ts (PPPTypes ts)
  , SOP.AllZip (SOP.LiftedCoercible (Term edsl) SOP.I) (PPPTypes ts) ts
  , SOP.All (IsPType edsl) (PPPTypes ts)
  ) =>
  PIsProduct (edsl :: PDSLKind) (ts :: [Type])
  where
  type PPPTypes ts :: [PType]
instance PIsProduct edsl '[] where
  type PPPTypes '[] = '[]
instance (IsPType edsl a, PIsProduct edsl ts) => PIsProduct edsl (Term edsl a : ts) where
  type PPPTypes (Term edsl a : ts) = a : PPPTypes ts

class
  ( SOP.AllZip2 (SOP.LiftedCoercible SOP.I (Term edsl)) tss (PSPTypes tss)
  , SOP.AllZip2 (SOP.LiftedCoercible (Term edsl) SOP.I) (PSPTypes tss) tss
  , SOP.SameShapeAs tss (PSPTypes tss)
  , SOP.SameShapeAs (PSPTypes tss) tss
  , SOP.AllZipF (SOP.AllZip (SOP.LiftedCoercible (Term edsl) SOP.I)) (PSPTypes tss) tss
  , SOP.AllZipF (SOP.AllZip (SOP.LiftedCoercible SOP.I (Term edsl))) tss (PSPTypes tss)
  , SOP.All2 (IsPType edsl) (PSPTypes tss)
  , SOP.SListI2 tss
  ) =>
  PIsSum (edsl :: PDSLKind) (tss :: [[Type]])
  where
  type PSPTypes (tss :: [[Type]]) :: [[PType]]
instance PIsSum edsl '[] where
  type PSPTypes '[] = '[]
instance (PIsProduct edsl ts, PIsSum edsl tss) => PIsSum edsl (ts : tss) where
  type PSPTypes (ts : tss) = PPPTypes ts : PSPTypes tss

class
  ( PGeneric a
  , PIsSum edsl (SOPG.GCode (PConcrete edsl a))
  , PReprSort a ~ PReprSOP
  ) =>
  PIsSOP (edsl :: PDSLKind) (a :: PType)
instance
  ( PGeneric a
  , PIsSum edsl (SOPG.GCode (PConcrete edsl a))
  , PReprSort a ~ PReprSOP
  ) =>
  PIsSOP (edsl :: PDSLKind) (a :: PType)

sopFrom ::
  forall edsl a.
  PIsSOP edsl a =>
  Proxy edsl ->
  Proxy a ->
  SOP.SOP SOP.I (SOPG.GCode (PConcrete edsl a)) ->
  SOP.SOP (Term edsl) (PSPTypes (SOPG.GCode (PConcrete edsl a)))
sopFrom _ _ = SOP.hcoerce

sopTo ::
  forall edsl a.
  PIsSOP edsl a =>
  Proxy edsl ->
  Proxy a ->
  SOP.SOP (Term edsl) (PSPTypes (SOPG.GCode (PConcrete edsl a))) ->
  SOP.SOP SOP.I (SOPG.GCode (PConcrete edsl a))
sopTo _ _ = SOP.hcoerce
```

specifically the "inner" types still don't leak to the class signature, which is nice.

**Comment by AriFordsham** [comment]
I see 'leaking' the signature as a feature, not a bug - those types are useful and not particulary private.

@re-xyr Would you mind to submit that as a PR against this branch so I can look at it properly?

**Comment by re-xyr** [comment]
I opened a PR. I don't really mind leaking the signatures or not, because I can't think of any significant downside to either. @L-as what do you think?

**Comment by AriFordsham** [comment]
Oh, I thought you got fundeps working. I specifically wrote this change in order to expose the inner types.

**Comment by re-xyr** [comment]
Well, I guess I'll try to make fundeps work tomorrow...

**Comment by AriFordsham** [comment]
I got as far as getting lost trying to define a pile of injective type families.

**Comment by L-as** [comment]
This is an improvement.

**Pull request merged**